[ENH] Compute fiber density and fiber spread from ODF using Bingham distributions

CHrlS98 opened the pull request:
As discussed with @villalonreina @RafaelNH @mdesco,
This is an attempt at migrating code from [`scilpy`](https://github.com/scilus/scilpy) for fitting Bingham distributions to ODF and computing fiber metrics such as the fiber density and fiber spread [1]. The idea was that this could serve as a basis for [GSoC 2023 Project 2: *Generalized along tract analysis of fiber orientation dispersion*](https://github.com/dipy/dipy/wiki/Google-Summer-of-Code-2023#project-2-generalized-along-tract-analysis-of-fiber-orientation-dispersion).

The PR is WIP. It still misses unit tests and stuff. I would mostly like some feedback at this point:
1. What would be the preferred data structure for representing Bingham distributions and their derived metrics. Right now I simply use python lists of tuples, but maybe we would like to use a numpy array like we do for tensors... It seems clearer as tuples because each parameter of the distribution is a separate output.
2. Do we want to support arrays of SF instead of a single SF at this point? For instance, given some ODF image, we could output a "Bingham image" like it's done in our `scilpy` script [here](https://github.com/scilus/scilpy/blob/master/scripts/scil_fit_bingham_to_fodf.py). Of course it would be cool, but maybe not needed in a first PR.
3. This does not feel like a `reconst` module. It's more of a voxel based analysis or something. Should this go in another module? Is it worth creating a new one?

Checklist
* [x] Write unit tests.
* [x] Support array of SF instead of single SF?

References
[1] Riffert TW, Schreiber J, Anwander A, Knösche TR. Beyond fractional anisotropy: Extraction of bundle-specific structural metrics from crossing fiber models. NeuroImage. 2014 Oct 15;100:176-91.

villalonreina commented:
Hi @CHrlS98, thanks for initiating this pull request. Regarding your questions:
1. We can keep it as lists and tuples for now, but at the end we must convert it to a numpy array to build the Nifti file for the full brain image.
2. We will add the feature to loop over the whole brain image, that is to do it over an array of ODFs (á la scilpy).
3. @RafaelNH and I discussed it and we think it should go into the `directions` module.

pep8speaks commented:
Hello @CHrlS98, Thank you for updating !

* In the file [`dipy/reconst/bingham.py`](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py):

> [Line 20:29](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py#L20): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 42:21](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py#L42): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 47:23](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py#L47): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 247:30](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py#L247): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 247:32](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py#L247): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 247:34](https://github.com/dipy/dipy/blob/82041cf2eb2f0f83e1c7d14d1533846800a04634/dipy/reconst/bingham.py#L247): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

villalonreina commented on dipy/reconst/bingham.py:
```diff
@@ -0,0 +1,256 @@
+from dipy.direction import peak_directions
+
+
+def bingham_to_sf(f0, k1, k2 , major_axis, minor_axis, vertices):
```
Can you change this function's name.? It's a bit confusing because SF is too close to FS which is the name of one of the output quantities. We suggest changing it to `bingham_odf`, because this function computes an ODF based on a Bingham function.

villalonreina commented on dipy/reconst/bingham.py:
```diff
@@ -0,0 +1,256 @@
+    return sf.T
+
+
+def bingham_fit_sf(sf, sphere, max_search_angle=15,
```
The same for this name, can you remove the 'sf' from this function name? this is for clarity so that it's not to be confused with the FS output parameter.

villalonreina commented on dipy/reconst/bingham.py:
```diff
@@ -0,0 +1,256 @@
+    return f0, k1, k2, mu1, mu2
+
+
+def bingham_to_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
```
Can you simplify this name to `bingham_fiber_density`?

villalonreina commented on dipy/reconst/bingham.py:
```diff
@@ -0,0 +1,256 @@
+    return fd
+
+
+def bingham_to_fiber_spread(bingham_fits, fd=None):
```
The same for this one, can you simplify it to `bingham_fiber_spread`?

villalonreina commented:
Hi @skoudoro. @CHrlS98 and @RafaelNH and I have been working on this PR. I am trying to do a pull request to Charles' PR but Charles' Dipy fork does not show up in Dipy's Pull request interface. See screenshot attached. Do you know what might be the issue causing this?
![Screen Shot 2024-02-26 at 9 09 33 AM](https://github.com/dipy/dipy/assets/11508365/8c5a4d83-83f8-436d-994b-fea8387e5563)

skoudoro commented:
Hi @villalonreina,

No idea, but when it does not show up, just go directly to his repo and it will show up: https://github.com/CHrlS98/dipy/pulls

skoudoro commented:
Also, @CHrlS98 @villalonreina,

Release in 2 weeks, I was hoping this will be in, but I am not sure it is ready. 
Please, Can you provide some feedback?

CHrlS98 commented:
@skoudoro
We are actively working on it. @villalonreina made some changes we will integrate to the pull request this week and then @RafaelNH will make a pass on it.

I think Rafael might have a better idea whether 2 weeks is a realistic timeline.

skoudoro commented:
Thank you for the feedback !

villalonreina commented:
Thanks @skoudoro. When I go to [https://github.com/CHrlS98/dipy/pulls](url) and I try to submit a PR, Github takes me out of the fork of @CHrlS98 and takes me to the Dipy's main site to make a PR directly to dipy/dipy, not CHrlS98/dipy. Anyway, I think that Charles found a way around. We will try to have this ready asap, hopefully for the next release.

RafaelNH commented:
I removed the WIP tag since we have a first version of everything

RafaelNH commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+           and their Application to the Healthy Ageing Brain (Doctoral thesis).
+           Downing College, University of Cambridge. doi: 10.17863/CAM.29356
+    """
+    if not (np.linalg.norm(vertices, axis=-1) == 1).any():
```
Perhaps we should consider removing this warning here given that this function is being used at a low level, e.g. this if statement is being repeatedly evaluated for all voxels in functions such as "bingham_from_odf" and "bingham_from_sh".

CHrlS98 replied:
I did like we discussed, that is, I added a `_bingham_to_odf` internal method in which the check is not done.

RafaelNH reviewed (changes requested):
Just two last comments at my side

RafaelNH commented on doc/examples/_valid_examples.toml:
```diff
@@ -111,6 +111,7 @@ files = [
     "reconst_qti.py",
     "reconst_qtiplus.py",
     "histology_resdnn.py",
+    "bingham_fit_fodf.py"
```
I forgot to add the new tutorial here - we have to do that.
Also I didn't delete the old example "bingham_fit_fodf.py". What should we do with it? Integrated it with the new tutorial? Delete it? This is something that we need to discuss.

skoudoro commented on dipy/direction/__init__.py:
```diff
@@ -4,3 +4,4 @@
 from .probabilistic_direction_getter import DeterministicMaximumDirectionGetter
 from .ptt_direction_getter import PTTDirectionGetter
 from .peaks import *
+from .bingham import *
```
`import *` to avoid. Please import only what you think necessary. 

I will create a PR to remove the one above (`.peaks import *`). I just saw it and this is the only one in the codebase

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+
+
+def _bingham_fit_peak(sf, peak, sphere, max_angle):
+    r"""
```
`r` is not necessary, to remove

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+    return f0, k1, k2, mu0, mu1, mu2
+
+
+def bingham_fit_odf(odf, sphere, npeaks=5, max_search_angle=6,
+                    min_sep_angle=60, rel_th=0.1):
```
We are slowly moving to keyword only so can you replace this by:  

```
def bingham_fit_odf(odf, sphere, *, npeaks=5, max_search_angle=6,
                    min_sep_angle=60, rel_th=0.1):
```

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+
+def bingham_fit_odf(odf, sphere, npeaks=5, max_search_angle=6,
+                    min_sep_angle=60, rel_th=0.1):
+    r"""
```
`r` not necessary

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+        The ODF function evaluated on the vertices of `sphere`
+    sphere: `Sphere` class instance
+        The Sphere providing the odf's discrete directions
+    npeak: int
```
`, optional`

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+    return fn.T
+
+
+def bingham_multi_voxel_odf(bingham_params, sphere, mask=None):
```
same comment as above some keyword-only

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+            and elements of Bingham dispersion minor axis (indexes 9-11).
+    sphere: `Sphere` class instance
+         The Sphere providing the odf's discrete directions
+    mask: ndarray
```
`, optional`

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+    return odf
+
+
+def bingham_fiber_density(bingham_params, mask=None, n_thetas=50, n_phis=100):
```
same comment as above for keyword-only

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,772 @@
+    """
+    # abs for twice the number of pts to fit
+    dot_prod = np.abs(sphere.vertices.dot(peak))
+    min_dot = cos(radians(max_angle))
```
to be consistent with all the codebase, I would recommend using numpy here (`np.cos`).  You use it later your code, so any reason to use the math module here ?

CHrlS98 replied:
Legacy code from scilpy. No there is no reason for not using numpy, I changed it.

skoudoro commented on doc/examples/bingham_fit_fodf.py:
```diff
@@ -0,0 +1,126 @@
+from dipy.viz import window, actor
+
+
+"""
```
old formatting style, Can you update this tutorial by following the new formatting style. Check your other example

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,304 @@
+FD_ODF_l2 = BinghamMetrics.fd[:, :, 0, 1]
+FD_total = BinghamMetrics.tfd[:, :, 0]
+
+fig1, ax = plt.subplots(1, 3, figsize=(16, 4))
+
+im0 = ax[0].imshow(FD_ODF_l1[:, -1:1:-1].T, vmin=0, vmax=2)
+ax[0].set_title('FD ODF lobe 1')
+
+im1 = ax[1].imshow(FD_ODF_l2[:, -1:1:-1].T, vmin=0, vmax=2)
+ax[1].set_title('FD ODF lobe 2')
+
+im2 = ax[2].imshow(FD_total[:, -1:1:-1].T, vmin=0, vmax=2)
+ax[2].set_title('FD total')
+
+fig1.colorbar(im0, ax=ax[0])
+fig1.colorbar(im1, ax=ax[1])
+fig1.colorbar(im2, ax=ax[2])
```
check in dipy.viz.plotting is there not a function that satisfy your needs. otherwise, create one or update one

CHrlS98 replied:
Created a new `image_mosaic` method in `viz.plotting`.

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,304 @@
+ODI1[FD_total < 0.5] = 0
+ODI2[FD_total < 0.5] = 0
+
+fig2, ax = plt.subplots(1, 3, figsize=(15, 5))
+
+im0 = ax[0].imshow(ODI1[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[0].set_title('ODI1 (lobe 1)')
+
+im1 = ax[1].imshow(ODI2[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[1].set_title('ODI2  (lobe 1)')
+
+im2 = ax[2].imshow(ODIt[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[2].set_title('ODIt  (lobe 1)')
+
+fig2.colorbar(im0, ax=ax[0])
+fig2.colorbar(im1, ax=ax[1])
+fig2.colorbar(im2, ax=ax[2])
```
same comment as above

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,304 @@
+ODI1[FD_total < 0.5] = 0
+ODI2[FD_total < 0.5] = 0
+
+fig3, ax = plt.subplots(1, 3, figsize=(15, 5))
+
+im0 = ax[0].imshow(ODI1[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[0].set_title('ODI1 (lobe 2)')
+
+im1 = ax[1].imshow(ODI2[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[1].set_title('ODI2  (lobe 2)')
+
+im2 = ax[2].imshow(ODIt[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[2].set_title('ODIt  (lobe 2)')
+
+fig3.colorbar(im0, ax=ax[0])
+fig3.colorbar(im1, ax=ax[1])
+fig3.colorbar(im2, ax=ax[2])
```
Same comment as above

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,304 @@
+ODI1[FD_total < 0.5] = 0
+ODI2[FD_total < 0.5] = 0
+
+fig4, ax = plt.subplots(1, 3, figsize=(15, 5))
+
+im0 = ax[0].imshow(ODI1[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[0].set_title('ODI1 (global)')
+
+im1 = ax[1].imshow(ODI2[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[1].set_title('ODI2  (global)')
+
+im2 = ax[2].imshow(ODIt[:, -1:1:-1].T, vmin=0, vmax=0.2)
+ax[2].set_title('ODIt  (global)')
+
+fig4.colorbar(im0, ax=ax[0])
+fig4.colorbar(im1, ax=ax[1])
+fig4.colorbar(im2, ax=ax[2])
```
same comment as above

skoudoro reviewed (commented):
Hi @CHrlS98,

I need to take time to go deeper in this code but thank you for this. 
See below, a very fast review for stuff that catch my eyes.

Also, there are many typo, please, look at codespell CI result to fix them.

CHrlS98 commented:
Thanks for the comments @skoudoro . We had a meeting with @villalonreina and @RafaelNH earlier today to discuss things we want to investigate or modify before merging. We'll get back to you for a second round of reviews once they are done.

CHrlS98 commented:
@villalonreina @RafaelNH 
So I worked on the code and here are a few comments we might want to discuss in a future meeting.

* __On the fiber density__
I changed the way the fiber density is estimated. Instead of creating a `(theta, phi)` grid with constant theta and phi, I use a DIPY repulsion sphere: this way, we are sure all points are at equal distance and cover the same surface area. The sphere can be passed as a parameter. I also added some tests for the new method.

* __On the maximum separation angle__
I tried lowering it using the tests like we discussed. It would seem there is a difference between `odf_to_bingham` and `_bingham_fit_peak` because for the former I can decrease the angle to 35 while for the later it won't work below 45.

* __`odf_to_bingham` and `bingham_from_odf`__
Both these methods exist, and from my understanding, `bingham_from_odf` takes a whole ODF volume as input whereas `odf_to_bingham` takes a single ODF as input. We might want to change the name because it is quite confusing.

CHrlS98 commented:
TODO
* [ ] T-design for default `bingham_fiber_density()`
* [ ] `peaks_direction_nl` or subdivide the sphere even more for `_bingham_fit_peak` and `odf_to_bingham` to give the same result in tests.
* [ ] always use multi-voxel methods for single and multi-voxel. Names `bingham_to_sf`, `sh_to_bingham`, `sf_to_bingham`.

CHrlS98 commented:
Hi @RafaelNH, is it possible t-design is slightly biased? I create my sphere this way (it doesn't matter if I subdivide it or not, I tried both) :
```
# default sphere is a subdivision of 45 points spherical t-design (11010 points)
verts = np.loadtxt(get_fnames('t-design'))
sphere = Sphere(xyz=verts).subdivide(4)
```

And then I added a test where I check if the integral of a Bingham distribution along the y-axis is equal to the integral along the x-axis:
```
# Rotating the Bingham distribution should not bias the FD estimate
xfits = [(f0_lobe1, k1, k2, axis0, axis1, axis2),
         (f0_lobe1, k1, k2, axis1, axis2, axis0)]
xpars = _convert_bingham_pars(xfits, 2)
xfd = bingham_fiber_density(xpars)

assert_almost_equal(xfd[0], xfd[1])
```

And sure it's close, but not even that close:
```
E           AssertionError: 
E           Arrays are not almost equal to 7 decimals
E            ACTUAL: 6.6747531900990253
E            DESIRED: 6.690992893100228
```
However when I use a unit icosahedron and subdivide it 5 times (10242 vertices), the test passes. I'll keep the unit icosahedron for now because it is better distributed than repulsion spheres, and add an option to control the level of subdivision.

Is the t-design a whole sphere or a hemisphere? Could it be the reason for this bias?

CHrlS98 commented on dipy/direction/tests/test_bingham.py:
```diff
@@ -0,0 +1,204 @@
+
+    # Test Bingham fit on full sampled GT Bingham function
+    odf_gt = _single_bingham_to_sf(f0, k1, k2, ma_axis, mi_axis, sphere.vertices)
+    a0, c1, c2, mu0, mu1, mu2 = _bingham_fit_peak(odf_gt, peak_dir, sphere, 45)
```
@RafaelNH If I replace `peak_dir` by `-peak_dir` I can lower the max angle to lesser than 45 and the test comparing `Mus` and Mus_ref` will still pass.

villalonreina commented:
@CHrlS98 @RafaelNH I made changes to the documentation, fixed typos and fixed references too. I made a PR to Charles' branch. Also, we should solve the angle (`max_search_angle`) question. I tested 6 degrees and 45 degrees in the example and the 45 degree option looks unrealistic. They both look very different when fitted on a brain image.

CHrlS98 commented:
@villalonreina
If we want to move forward with the PR, should we just make `max_search_angle` a mandatory argument and let the user play with it until we figure it out?

villalonreina commented:
Hi @CHrlS98 , yes I do think so as well, it should be a mandatory input. I was trying to run validation metrics in July and we still have to figure out what the best option would be. What do you think @RafaelNH?

skoudoro commented:
Hello @CHrlS98, @villalonreina and @RafaelNH,

By curiosity, what is the status of this work?

villalonreina commented:
> Hello @CHrlS98, @villalonreina and @RafaelNH,
> 
> By curiosity, what is the status of this work?

Hi @skoudoro. Sorry for the delay. I think that the PR is pretty much ready. We have problems though with one of the inputs to the bingham function, i.e. the search angle for the FOD lobe. We decided to include it as a compulsory argument for the end user, but we don't have a clear recommnedation yet for what it should be. We will all meet to discuss about this issue. I think that we need testing and validation. Not sure if the PR can be merged to master before doing that? What do you think? @RafaelNH @CHrlS98

CHrlS98 commented:
@villalonreina @skoudoro I think all the new functions are tested and documented, so if we just change `max_search_angle` to make it mandatory I think it could be merged. I don't know if I'll have time for this before October however.

@RafaelNH Could you go for a second round of review when you have time?

Thanks!

CHrlS98 commented:
@villalonreina @RafaelNH @skoudoro
`max_search_angle` is now a mandatory argument. I would be comfortable merging the PR even without extensive validation of the `max_search_angle` argument. We have time for recommendations later, and at least the code will already be available to our users.

Review when you have some time, thank you!

villalonreina commented:
Hi @CHrlS98. Thanks so much for the update. Yes, I agree with this step. `max_search_angle` should be mandatory even though we still don't have a clear recommendation of what it should be. Anything from 6 to 45 grades. @RafaelNH and I will meet in a few weeks and will discuss about the issue to try to resolve it, hopefully with some validation (simulations or phantoms?). We'll keep you posted. @skoudoro is up to you and the rest of the team to decide whether it should be merged like this or not. Thanks.

skoudoro reviewed (commented):
Hi @CHrlS98,

Thank you for this clean code. Overall, it looks good, you just have to update the citing system since we changed and use a more centralized way for citation. I am ok to merge as it is after your update.

I agree that the recommendation can be done  later.

Thanks to @villalonreina and @RafaelNH  for the reviews!

CHrlS98 commented on dipy/direction/tests/test_bingham.py:
```diff
@@ -0,0 +1,208 @@
+
+    bim_odf = sf_to_bingham(odf, sphere, npeaks=2, max_search_angle=45)
+    sh = sf_to_sh(odf, sphere, sh_order_max=16)
+    bim_sh = sh_to_bingham(sh, sphere, 16, npeaks=2, max_search_angle=45)
```
I instead added an argument `legacy` to `sh_to_bingham` and set it to false in tests to avoid the warning.

CHrlS98 commented:
@skoudoro

Thank you for the feedback. I answered your review comments.

Let me know if anything else needs changing.

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,315 @@
+
+###############################################################################
+# Step 1. ODF estimation
+# =============================================================================
```
I mean, we follow the rst rules: https://docutils.sourceforge.io/docs/user/rst/quickstart.html#sections

so, it should be this (underline same size as the title):
```rst
###############################################################################
# Step 1. ODF estimation
# ======================
```

Can you update everywhere

CHrlS98 replied:
Alright done!

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,315 @@
+#
+#
+# Step 2. Bingham fitting and Metrics
+# =============================================================================
```
same comment as above

skoudoro commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,315 @@
+
+###############################################################################
+# Step 3. Bingham Metrics
+# =============================================================================
```
same comment as above

skoudoro reviewed (commented):
Thank you for the update @CHrlS98,

See below the answer to your question

villalonreina commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,787 @@
+    max_search_angle: float.
+        Maximum angle between a peak and its neighbour directions
+        for fitting the Bingham distribution. Although they suggest 6 degrees
+        in [1]_, tests show that a value around 45 degrees is more stable.
```
@CHrlS98 @skoudoro I think this is the input to make mandatory instead of optional, is that right?

CHrlS98 replied:
Yes, I removed the default value. It must be specified by the user.

skoudoro commented:
Hi @CHrlS98 ,

Thank you for the  lastupdate. 

Code format CI is still failing as you can see here: https://github.com/dipy/dipy/actions/runs/11389486311/job/31688883284?pr=2826#step:4:137

A good way to check that locally is to use pre-commit:
```
pip install pre-commit 
pre-commit install
pre-commit run --files your_files.py
```

Also, build_docs CI's is failing also. here are the warnings, but you can test it locally by doing in the `docs`
 folder: `make - C . html-no-examples` (you might have to update the requirements, check doc-requirements.txt)

```python
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/numpydoc/docscrape.py:421: UserWarning: Unknown section Return in the docstring of sf_to_bingham in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py.
  self[section] = content
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/numpydoc/docscrape.py:421: UserWarning: Unknown section Return in the docstring of sh_to_bingham in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py.
  self[section] = content
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/numpydoc/docscrape.py:384: UserWarning: potentially wrong underline length... 
Note: 
---- in 
Overall concentration parameters for an ODF peak.
... in the docstring of kappa_total_lobe in /opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py.
  if not self._is_at_section():
```

CHrlS98 commented:
@skoudoro 
Remaining doc error seems to be
```
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py:docstring of dipy.direction.bingham.BinghamMetrics.kappa_total_lobe:4: ERROR: Too many autonumbered footnote references: only 0 corresponding footnotes available. [docutils]
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py:docstring of dipy.direction.bingham.BinghamMetrics.kappa_total_lobe:4: ERROR: Unknown target name: "footcite-tariq2016". [docutils]
/opt/hostedtoolcache/Python/3.11.10/x64/lib/python3.11/site-packages/dipy/direction/bingham.py:docstring of dipy.direction.bingham.BinghamMetrics.odi_total_lobe:2: ERROR: Unknown target name: "footcite-tariq2016". [docutils]
```

I can't figure out why this particular reference won't work. Would you happen to have a solution?

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,790 @@
+
+    @auto_attr
+    def odi_total_lobe(self):
+        """Overall Orientation Dispersion Index (ODI) computed for an
+        ODF lobe from the overall concentration parameter (k_total).
+        Defined by equation 20 in :footcite:p:`Tariq2016`.
```
I believe the reason is `footcite` in a short description.  it can not be in a short description. so splitting short and long description like below should work. (I did not try, I can test this later tonight)

```python
"""Return Orientation Dispersion Index (ODI) for an ODF lobe.

The ODF lobe is from the overall concentration parameter (k_total).Defined by equation 20 in :footcite:p:`Tariq2016`.

References
----------
.. footbibliography::
"""
```

CHrlS98 replied:
Thanks, it appears to have worked.

skoudoro commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,790 @@
+        """Overall concentration parameters for an ODF peak.
+
+        The overall (combined) concentration parameters for each lobe is
+        defined by equation 19 in :footcite:p:`Tariq2016` as
```
I believe it is good but disturbed by the other one. See my other comment.

skoudoro commented:
Hi @arokem,

Can you have a quick look at this PR before going ahead for merging?  another pair of eye is needed 😅 
@Atharva-Shah-2298 have checked it also. 

Thank you in advance

arokem commented:
Will do. Give me a day or two

skoudoro commented:
Hi @arokem,

Do you have any update on this? Can we go ahead ? Thank you for your feedback

arokem commented:
Sorry - I got distracted by many other things... 

Right off the bat, an answer to a question that was asked at the very top: I do think this should go into the reconst module. Seems unjustified to start a new module for this.

arokem commented:
Even the example (looks great!) is called reconst_bingham, suggesting that this is the module to look into.

arokem commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,792 @@
+        approximation will be, at the cost of longer execution times. The
+        default results in a sphere of 10242 points.
+    mask: ndarray, optional
+        Map marking the coordinates in the data that should be analyzed
```
```suggestion
        Map marking the coordinates in the data that should be analyzed.
        Default (None) means all voxels in the volume will be analyzed.
```

arokem commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,792 @@
+    Parameters
+    ----------
+    bmetric: ndarray(..., nl)
+        An arbitrary metric with values for nl ODF lobes.
```
```suggestion
        Any metric with values for nl ODF lobes.
```

arokem commented on dipy/direction/bingham.py:
```diff
@@ -0,0 +1,792 @@
+    if mask is None:
+        mask = np.ones(shape)
+
+    # Bingham parameters saved in an ndarray with shape:
```
```suggestion
    # Bingham parameters stored in an ndarray with shape:
```

arokem reviewed (commented):
This is great work! I look forward to using it. 

Other than my comment about whether to include this in the `reconst` module (I think so, but not adamant), I had just a few small comments.

skoudoro commented:
Thanks @arokem!

Good catch! +1 for having this in `reconst` module

skoudoro commented:
Hi @CHrlS98,

it would be great if you could address ariel comment today or tomorrow morning 😅

skoudoro reviewed (approved):
Thank you for this great work @CHrlS98!

Thank you for all the reviews @RafaelNH, @villalonreina, @arokem

merging
